src/mds/Locker.cc

```diff
@@ -935,6 +938,27 @@ void Locker::eval_gather(SimpleLock *lock, bool first, bool *pneed_issue, MDSCon
 	    mds->send_message_mds(make_message<MLock>(lock, LOCK_AC_LOCKACK, mds->get_nodeid(), data), auth);
 	    (static_cast<ScatterLock *>(lock))->start_flush();
 	    // we'll get an AC_LOCKFLUSHED to complete
+
+            // In the following scenario, the current lock's dirty state could be lost:
+            // 1. current lock's state is LOCK_LOCK;
+            // 2. auth send LOCK_AC_MIX, and the current lock do finish_scatter_update, which
+            //    wrlock_force the current lock;
+            // 3. auth send LOCK_AC_LOCK, and would be forced to wait, since the current lock
+            //    is wrlocked;
+            // 4. the child inode want to pop rstat up to the current lock's parent, which would
+            //    also be forced to wait since the current lock state is LOCK_MIX_LOCK; the current
+            //    lock would be marked dirty, and the child inode would be attached to the corresponding
+            //    dir's dirty_rstat_inodes;
+            // 5. the previous finish_scatter_update finishes, the codes reaches here and "start_flush"
+            //    the current lock, which would clear the dirty flag of the current lock.
+            // As the CInode::encode_lock_state method doesn't pop dirty_rstat_inodes' rstat to the corresponding
+            // dir, step 5 actually lead to the lost of the dirty state marked by step 4.
```

Comment:
maybe this is an indication that we should use separate flag for dirty rstat and direcotory has children with dirty rstat

Reply:
Yes, we can try that:-)

Reply:
@ukernel Hi, it seems that the dirty state is now relied upon by many other function/modules, splitting it would involve a lot of modification. On the other hand, the problem is caused by the clearing of the dirty state not taking into account cases that there are children with dirty rstat, so I think it should be enough to make ScatterLock::clear_dirty to test whether there is such children and not clear the dirty state if it does. Am I right? Thanks:-)

Reply:
@ukernel And if the dirty state flag does need to be splitted, i think we can put that in future patches. After all, we can't solve everything within a single PR, am I right? Thanks:-)

Reply:
@ukernel Hi, I've seperated the commit for solving this problem into another PR: [29721](https://github.com/ceph/ceph/pull/29721) since it seems that this bug should be fixed  for systems without rstat flush funcion. Please take a look:-) Thanks:-)